### PR TITLE
feat/complex-x-ts-types - Adds supports for using x-ts-type in `anyOf`, `allOf`, `oneOf`, and `not` keywords

### PR DESCRIPTION
**Pull request by mschnee**
Partially implements: #4300 
See: #4406

This has been split from #4406 as a more manageable, smaller PR.

## Added
This PR adds support for using `x-ts-type` schema bodies the value of `anyOf`, `allOf`, `oneOf`, and `not` schema object keys, as described by the [OpenAPI Schema Object specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#schema-object).  In this example `ResponseObject`:
```ts
{
  description: 'Path Response Description',
  content: {
    'application/json': {
      schema: {
        anyOf: [
          { 'x-ts-type': CouldBeThisModel },
          { 'x-ts-type': OrThisModel } 
        ],
        not: { 'x-ts-type': ButNotThisModel }
      }
    }
  }
}
```

## Checklist
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] (n/a) Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: packages/openapi-v3/src/controller-spec.ts [hunk 1]
```diff
@@ -248,24 +251,50 @@ function processSchemaExtensions(
...
+  if (schema.not) {
+    processSchemaExtensions(spec, schema.not);
+  }
+
+  /**
+   *  check for schema.allOf, schema.oneOf, schema.anyOf arrays first.
+   *  You cannot provide BOTH a defnintion AND one of these keywords.
+   */
+  const hasOwn = (prop: string) =>
+    Object.prototype.hasOwnProperty.call(schema, prop);
```
Comment (raymondfeng) [inline packages/openapi-v3/src/controller-spec.ts]: Can it be just the following?
```ts
const hasOwn = (prop: string) => schema.hasOwnProperty(prop);
```

**Comment by mschnee** [comment]
NOTE I've rebased and force-pushed my branch.  I hope that addresses the travis-commitlint build errors, as I think that may have come in from pulling upstream.

File: packages/openapi-v3/src/controller-spec.ts (hunk 1 above)
Comment (mschnee) [reply packages/openapi-v3/src/controller-spec.ts]: Sure!  Sorry about that, I'm so used to that being a lint-error/typescript-warning these days.

**Comment by dhmlau** [comment]
FYI - the drop in test coverage is in the `examples` package which I don't think it's related to the changes in this PR.

**Comment by mschnee** [comment]
```
$ source cloudant-config.sh && npm run mocha
/home/travis/.travis/functions: line 109: cloudant-config.sh: No such file or directory
```
:/

**Comment by raymondfeng** [comment]
@mschnee Please ignore the cloudant issue.

**Comment by mschnee** [comment]
I will update the documentation referring to `x-ts-type` to be more explicit about anyOf/allOf/oneOf/not tomorrow.

**Comment by mschnee** [comment]
Document updated :)

**Comment by raymondfeng** [comment]
@mschnee Please rebase your branch to latest master to pass CI tests.

**Comment by raymondfeng** [comment]
@mschnee Please follows instructions at https://loopback.io/doc/en/lb4/submitting_a_pr.html#6-rebasing for rebasing.

File: package.json [hunk 2]
```diff
@@ -42,7 +43,7 @@
     "tsdocs": "lerna run --scope @loopback/tsdocs build:tsdocs",
     "coverage:ci": "node packages/build/bin/run-nyc report --reporter=text-lcov | coveralls",
     "precoverage": "npm test",
-    "coverage": "open coverage/index.html",
+    "coverage": "open-cli coverage/index.html",
```
Comment (raymondfeng) [inline package.json]: Can we move this change to a separate PR?

Comment (mschnee) [reply package.json]: Done!

**Comment by mschnee** [comment]
Rebase complete! :)

**Comment by raymondfeng** [comment]
@mschnee Thank you for cleaning up the PR. One last step - please squash two commits into one per instructions at https://loopback.io/doc/en/lb4/submitting_a_pr.html#10-final-rebase-and-squashing-of-commits.

**Comment by mschnee** [comment]
Rebased off of master this morning again and pushed :)
